tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php

```diff
@@ -863,6 +864,84 @@ private function skipOci8AndMysqli()
             $this->markTestSkipped('Mysqli driver dont support this feature.');
         }
     }
+
+    public function testIteratorCallsFetchTheExpectedNumberOfTimes()
+    {
+        $sql = 'SELECT test_int, test_string FROM fetch_table';
+        $originalStmt = $this->_conn->prepare($sql);
+
+        // Mock the connection
+        $connectionMock = $this->getMockBuilder('Doctrine\\DBAL\\Driver\\Connection')
+            ->setMethods(['prepare', 'query'])
+            ->getMockForAbstractClass();
+
+        $stmtSpy = new StatementSpy($originalStmt);
+
+        // Return our mocked statement from the query
+        $connectionMock->expects($this->any())
+            ->method('prepare')
+            ->will($this->returnValue($stmtSpy));
+
+        $connectionMock->expects($this->any())
+            ->method('query')
+            ->will($this->returnCallback(function () use ($stmtSpy) {
+                $stmtSpy->execute();
+                return $stmtSpy;
+            }));
+
+        $stmt = $connectionMock->query($sql);
+        $stmt->setFetchMode(\PDO::FETCH_ASSOC);
+
+        // Assert no fetch calls thus far
+        self::assertEquals(0, $stmtSpy->getFetchCalls());
+
+        $i = 0;
+        foreach ($stmt as $row) {
+            $i++;
+        }
+
+        // Assert expected number of fetches
+        $this->assertGreaterThan(0, $stmtSpy->getFetchCalls());
+        $this->assertEquals($i, $stmtSpy->getFetchCalls());
```

Comment:
It doesn't guarantee that `fetch()` is called once per step. It would also pass if the code did `while (fetch()); return new ArrayIterator($rows)`;